segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -308,23 +336,76 @@ private boolean processWrites(List<Write> toExecute) {
-            return Collections.emptyList();
+    private boolean handleClosedLedgers(List<Write> writes) {
+        if (writes.size() == 0 || !writes.get(0).getWriteLedger().ledger.isClosed()) {
+            // Nothing to do. We only need to check the first write since, if a Write failed with LedgerClosed, then the
+            // first write must have failed for that reason (a Ledger is closed implies all ledgers before it are closed too).
+            return false;
         }
 
-        // Calculate how much estimated space there is in the current ledger.
-        final long maxTotalSize = this.config.getBkLedgerMaxSize() - getWriteLedger().ledger.getLength();
+        WriteLedger currentLedger = getWriteLedger();
+        Map<Long, Long> lastAddsConfirmed = new HashMap<>();
+        boolean anythingChanged = false;
+        for (Write w : writes) {
+            if (w.isDone() || !w.getWriteLedger().ledger.isClosed()) {
+                continue;
+            }
 
-        // Get the writes to execute from the queue.
-        return this.writes.getWritesToExecute(maxTotalSize);
+            // Write likely failed because of LedgerClosedException. Need to check the LastAddConfirmed for each
+            // involved Ledger and see if the write actually made it through or not.
+            long lac = tryGetLastAddConfirmed(w.getWriteLedger(), lastAddsConfirmed);
+            if (w.getEntryId() >= 0 && w.getEntryId() <= lac) {
+                // Write was actually successful. Complete it and move on.
+                w.complete();
+                anythingChanged = true;
+            } else if (currentLedger.ledger.getId() != w.getWriteLedger().ledger.getId()) {
+                // Current ledger has changed; attempt to write to the new one.
+                w.setWriteLedger(currentLedger);
+                anythingChanged = true;
+            }
+        }
+
+        return anythingChanged;
+    }
+
+    /**
+     * Reliably gets the LastAddConfirmed for the WriteLedger
+     *
+     * @param writeLedger       The WriteLedger to query.
+     * @param lastAddsConfirmed A Map of LedgerIds to LastAddConfirmed for each known ledger id. This is used as a cache
+     *                          and will be updated if necessary.
+     * @return The LastAddConfirmed for the WriteLedger.
+     */
+    @SneakyThrows(DurableDataLogException.class)
+    private long tryGetLastAddConfirmed(WriteLedger writeLedger, Map<Long, Long> lastAddsConfirmed) {
+        long ledgerId = writeLedger.ledger.getId();
+        long lac = lastAddsConfirmed.getOrDefault(ledgerId, -1L);
+        if (lac < 0) {
+            if (writeLedger.isRolledOver()) {
+                // This close was not due to failure, rather a rollover - hence lastAddConfirmed can be relied upon.
+                lac = writeLedger.ledger.getLastAddConfirmed();
+            } else {
+                // Ledger got closed. This could be due to some external factor, and lastAddConfirmed can't be relied upon.
```

Comment:
One situation in which the ledger can get closed is when the container ownership changes, yes? What happens in this case?